ddlib/ddlib/util.py

```diff
@@ -371,3 +379,45 @@ def tsv_extractor(generator):
   for row in parser.parse_stdin():
     for out_row in generator(**row._asdict()):
       printer.write(out_row)
+
+
+def tsj_extractor(generator):
+  """
+  When a generator function is decorated with this (i.e., @tsj_extractor
+  preceding the def line), each standard input line is parsed as
+  tab-separated JSON (TSJ) values, then the function is applied to the parsed
+  array to generate output rows, and each output row expected to be an array
+  is formatted as TSJ.
+  """
+  reload(sys).setdefaultencoding("utf8")  # to avoid UnicodeEncodeError of JSON values during conversion by str()
+
+  input_format, output_format = get_generator_format(generator)
+  input_names  = [name for name,t in input_format]
+  num_input_values = len(input_format)
+  num_input_splits = num_input_values - 1
+  num_output_values = len(output_format)
+
+  def parse_json(column_index, json_value):
```

Comment:
You mean `parse_json`?  It's just a `json.loads` with context-dependent error message translation.. that I doubt would be reusable or worth testing.  It's just to make the for loop body look cleaner with list comprehension syntax.  If the nested `def`s disturbs you in general, I can inline the code and make the following for loop longer.  I also hate `def`s but Python's dichotomy btwn statements and expressions forces you to use them.